src/cargo/util/lints.rs

```diff
@@ -0,0 +1,241 @@
+            let array = item.as_array().unwrap();
+            let next = iter.next().unwrap();
+            return array.iter().find_map(|item| {
+                if next == &item.to_string() {
+                    item.span()
+                } else {
+                    None
+                }
+            });
+        }
+    }
+    None
+}
+
+fn manifest_path(path: &Path, gctx: &GlobalContext) -> String {
+    diff_paths(path, gctx.cwd())
+        .unwrap_or_else(|| path.to_path_buf())
+        .display()
+        .to_string()
+}
+
+#[derive(Copy, Clone, Debug)]
+pub struct LintGroup {
+    pub name: &'static str,
+    pub default_level: LintLevel,
+    pub desc: &'static str,
+    pub edition_lint_opts: Option<(Edition, LintLevel)>,
+}
+
+const RUST_2024_COMPATIBILITY: LintGroup = LintGroup {
+    name: "rust-2024-compatibility",
+    default_level: LintLevel::Allow,
+    desc: "warn about compatibility with Rust 2024",
+    edition_lint_opts: Some((Edition::Edition2024, LintLevel::Deny)),
+};
+
+#[derive(Copy, Clone, Debug)]
+pub struct Lint {
+    pub name: &'static str,
+    pub desc: &'static str,
+    pub groups: &'static [LintGroup],
+    pub default_level: LintLevel,
+    pub edition_lint_opts: Option<(Edition, LintLevel)>,
+}
+
+impl Lint {
+    pub fn level(&self, lints: &TomlToolLints, edition: Edition) -> LintLevel {
+        let level = self
+            .groups
+            .iter()
+            .map(|g| g.name)
+            .chain(std::iter::once(self.name))
+            .filter_map(|n| lints.get(n).map(|l| (n, l)))
+            .max_by_key(|(n, l)| (l.priority(), std::cmp::Reverse(*n)));
+
+        match level {
+            Some((_, toml_lint)) => toml_lint.level().into(),
+            None => {
+                if let Some((lint_edition, lint_level)) = self.edition_lint_opts {
+                    if edition >= lint_edition {
+                        lint_level
+                    } else {
+                        IMPLICIT_FEATURES.default_level
+                    }
+                } else {
+                    IMPLICIT_FEATURES.default_level
+                }
+            }
+        }
+    }
+}
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum LintLevel {
+    Allow,
+    Warn,
+    Deny,
+    Forbid,
+}
+
+impl LintLevel {
+    pub fn to_diagnostic_level(self) -> Level {
+        match self {
+            LintLevel::Allow => Level::Note,
+            LintLevel::Warn => Level::Warning,
+            LintLevel::Deny => Level::Error,
+            LintLevel::Forbid => Level::Error,
+        }
+    }
+}
+
+impl Into<LintLevel> for TomlLintLevel {
+    fn into(self) -> LintLevel {
+        match self {
+            TomlLintLevel::Allow => LintLevel::Allow,
+            TomlLintLevel::Warn => LintLevel::Warn,
+            TomlLintLevel::Deny => LintLevel::Deny,
+            TomlLintLevel::Forbid => LintLevel::Forbid,
+        }
+    }
+}
+
+const IMPLICIT_FEATURES: Lint = Lint {
```

Comment:
Could we have a brief doc comment with examples explaining this lint?
(And later we can parse it and generate a website 🎉)

Reply:
Please let me know if you like the doc comment. I pulled it from the summary of [RFC #3491](https://rust-lang.github.io/rfcs/3491-remove-implicit-features.html).